### PR TITLE
Separate etcd certificates

### LINKED ISSUE #3091: etcd cipher suites
## Feature Request

Limit cipher suites in etcd to prevent TLS downgrade attacks

### LINKED ISSUE #3092: etcd client certs
## Feature Request

Generate and use etcd client certs when connecting as client to etcd instead of using peer certs as client certs.

### PR DESCRIPTION
**Pull request by sergelogvinov**
# Pull Request

Fixes #3092
Fixes #3091

## Acceptance

Please use the following checklist:

- [ ] you linked an issue (if applicable)
- [ ] you included tests (if applicable)
- [x] you ran conformance (`make conformance`)
- [x] you formatted your code (`make fmt`)
- [x] you linted your code (`make lint`)
- [ ] you generated documentation (`make docs`)
- [ ] you ran unit-tests (`make unit-tests`)

> See `make help` for a description of the available targets.

---
This change is [Reviewable](https://reviewable.io/reviews/talos-systems/talos/3777)

### CODE REVIEW

File: pkg/resources/secrets/etcd.go [hunk 1]
```diff
@@ -26,7 +26,9 @@ type Etcd struct {
 
 // EtcdCertsSpec describes etcd certs secrets.
 type EtcdCertsSpec struct {
-	EtcdPeer *x509.PEMEncodedCertificateAndKey `yaml:"etcdPeer"`
+	EtcdPeer      *x509.PEMEncodedCertificateAndKey `yaml:"etcdPeer"`
```
Comment (smira) [inline pkg/resources/secrets/etcd.go:29]: do we need `EtcdPeer` and `EtcdClient` here?

File: internal/app/machined/pkg/system/services/etcd.go [hunk 2]
```diff
@@ -341,13 +356,14 @@ func (e *Etcd) argsForInit(ctx context.Context, r runtime.Runtime) error {
 		"data-dir":              constants.EtcdDataPath,
 		"listen-peer-urls":      "https://" + net.FormatAddress(listenAddress) + ":2380",
 		"listen-client-urls":    "https://" + net.FormatAddress(listenAddress) + ":2379",
-		"cert-file":             constants.KubernetesEtcdPeerCert,
-		"key-file":              constants.KubernetesEtcdPeerKey,
+		"client-cert-auth":      "true",
+		"cert-file":             constants.KubernetesEtcdClientCert,
```
Comment (smira) [inline internal/app/machined/pkg/system/services/etcd.go:360]: this feels wrong: we should use server cert here (not client cert, and not peer cert)

File: pkg/resources/secrets/etcd.go (hunk 1 above)
Comment (sergelogvinov) [reply pkg/resources/secrets/etcd.go:29]: get etcdsecrets.secrets.talos.dev -oyaml -- looks nice
And you can check all certs. 

https://github.com/talos-systems/talos/pull/3777/files#diff-f028d821b1f248fa75f380f9a91477df94583726a996ead5f16b089ec24baf04R130 here updateSecrets func recreate the certs

File: internal/app/machined/pkg/system/services/etcd.go (hunk 2 above)
Comment (sergelogvinov) [reply internal/app/machined/pkg/system/services/etcd.go:360]: May be I need to rename it.
EtcdClientCert - this is a server certificate for client connections.

Comment (smira) [reply internal/app/machined/pkg/system/services/etcd.go:360]: looks like we have a mixed client/server cert there, that's why I asked

Comment (sergelogvinov) [reply internal/app/machined/pkg/system/services/etcd.go]: renamed EtcdClientCert -> EtcdCert

File: internal/app/machined/pkg/system/services/etcd.go [hunk 3]
```diff
@@ -147,6 +147,8 @@ func (e *Etcd) Runner(r runtime.Runtime) (runner.Runner, error) {
 		env = append(env, "ETCD_UNSUPPORTED_ARCH=arm64")
 	}
 
+	env = append(env, "ETCD_CIPHER_SUITES=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305") //nolint:lll
```
Comment (smira) [inline internal/app/machined/pkg/system/services/etcd.go:150]: what is the effect of this line on etcd security/performance?

File: go.sum [hunk 4]
```diff
@@ -1113,6 +1113,10 @@ github.com/syndtr/gocapability v0.0.0-20180916011248-d98352740cb2/go.mod h1:hkRG
 github.com/syndtr/gocapability v0.0.0-20200815063812-42c35b437635/go.mod h1:hkRG7XYTFWNJGYcbNJQlaLq0fg1yr4J4t/NcTQtrfww=
 github.com/talos-systems/crypto v0.3.1-0.20210615131117-6bc5bb50c527 h1:Ut3jI/0FPSyH0QpycKFKAueWKjzQ00OFmkeFG+0uO1A=
 github.com/talos-systems/crypto v0.3.1-0.20210615131117-6bc5bb50c527/go.mod h1:xaNCB2/Bxaj+qrkdeodhRv5eKQVvKOGBBMj58MrIPY8=
+github.com/talos-systems/crypto v0.2.1-0.20210601174604-cd18ef62eb9f h1:Xk3zeUZPhvEl9Vs4PlYBohin3QZmizA/YR4URKEyULY=
```
Comment (smira) [inline go.sum:1116]: feels like `go mod tidy` should clean these up

File: pkg/machinery/go.sum [hunk 5]
```diff
@@ -146,6 +146,10 @@ github.com/stretchr/testify v1.7.0 h1:nwc3DEeHmmLAfoZucVR881uASk0Mfjw8xYJ99tb5Cc
 github.com/stretchr/testify v1.7.0/go.mod h1:6Fq8oRcR53rry900zMqJjRRixrwX3KX962/h/Wwjteg=
 github.com/talos-systems/crypto v0.3.1-0.20210615131117-6bc5bb50c527 h1:Ut3jI/0FPSyH0QpycKFKAueWKjzQ00OFmkeFG+0uO1A=
 github.com/talos-systems/crypto v0.3.1-0.20210615131117-6bc5bb50c527/go.mod h1:xaNCB2/Bxaj+qrkdeodhRv5eKQVvKOGBBMj58MrIPY8=
+github.com/talos-systems/crypto v0.2.1-0.20210601174604-cd18ef62eb9f h1:Xk3zeUZPhvEl9Vs4PlYBohin3QZmizA/YR4URKEyULY=
```
Comment (smira) [inline pkg/machinery/go.sum:149]: same as above (go mod tidy)?

File: internal/pkg/etcd/certs.go [hunk 6]
```diff
@@ -57,3 +76,59 @@ func GeneratePeerCert(etcdCA *x509.PEMEncodedCertificateAndKey) (*x509.PEMEncode
...
+//nolint:dupl
+func GenerateClientCert(etcdCA *x509.PEMEncodedCertificateAndKey) (*x509.PEMEncodedCertificateAndKey, error) {
+	opts, err := NewCommonOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	opts = append(opts,
+		x509.ExtKeyUsage([]stdlibx509.ExtKeyUsage{
+			stdlibx509.ExtKeyUsageServerAuth,
```
Comment (smira) [inline internal/pkg/etcd/certs.go:91]: should we have both client and server cert to be same cert? or should we split client and server certs? 

not sure what is the best practice for etcd

File: internal/app/machined/pkg/system/services/etcd.go (hunk 3 above)
Comment (smira) [reply internal/app/machined/pkg/system/services/etcd.go:150]: ref (one of them): https://github.com/etcd-io/etcd/issues/10304

Comment (sergelogvinov) [reply internal/app/machined/pkg/system/services/etcd.go:150]: * AES 128/256 with RSA/ECDSA for CPU with AES-IN
* chacha-poly with RSA/ECDSA for ARM

https://blog.cloudflare.com/do-the-chacha-better-mobile-performance-with-cryptography/

etcd cluster works inside the tales cluster, and it does not support to connect external etcd nodes. We can set less choice but with better performance.

File: internal/pkg/etcd/certs.go (hunk 6 above)
Comment (sergelogvinov) [reply internal/pkg/etcd/certs.go:91]: definitely yes. 

But etcd uses this cert to check itself.

* https://kubernetes.io/docs/setup/best-practices/certificates/ uses both (clientAuth, serverAuth)
* https://github.com/kubernetes-sigs/kubespray/blob/master/roles/etcd/templates/openssl.conf.j2 uses both (clientAuth, serverAuth)

I've add talos etcd client cert, now we have 4 certs for etcd cluster.

**Comment by smira** [comment]
:+1: this looks good to me, I want to do a little bit of optimization to skip generating etcd certs twice in the controller and service code. 

I will push to this PR resolving conflicts.

**Review by smira — APPROVED** [review]

**Pull request merged**